apps/consulting/posts/improvements-to-bsr-matrix-multiplication-in-pytorch.md

```diff
@@ -0,0 +1,292 @@
+sparse argument were converted to dense, or all implicit zeros were materialized the result should
+be equivalent. The DSD pattern can be used to compute linear layers with sparse weights. This is an
+function which is semantically equivalent to [`torch.nn.functional.linear`][pytorch-docs-linear]
+
+```python
+def linear(x, W, b=None):
+    tmp = (W @ x.transpose(-2, -1)).transpose(-2, -1)
+    if b is not None:
+        return tmp + b
+    else:
+        return tmp
+```
+
+In this case the weights, `W` can be stored using BSR layout, provided the DSD matrix multiply is
+supported. Note that linear is defined by `x@W^T` and above we have calculated the transpose of this
+expression.
+
+In contrast the SDD pattern does is not semantically equivalent under conversion to dense layout. In
+this case the sparse tensor updated with the result of `A@B` also acts as a mask.  In other words
+`C` will have the same sparsity pattern and that pattern is imposed on the product `A@B`.  In
+PyTorch, we call this operation [`torch.sparse.sampled_addmm`][pytorch-docs-sampled-addmm].  This
+operation is also relevant for transformer models.  It can be used to compute masked gradients for
+terms like `linear` where during training we want to ensure the sparsity pattern of `W` is preserved.
+
+There exist BLAS-like libraries supporting sparse layouts, for example NVIDIA's
+[cuSPARSE][cusparse-docs]. These options are lacking some important features that are needed to make
+it practical to use sparse layouts in machine learning applications. First, they typically perform
+poorly compared to dense matrix multiply except when the sparsity is very high (<10% of the elements
+are stored). Second, many of the operations are missing support for half precision data types, which
+is a feature used frequently in learning applications.
+
+In order to overcome the limitations, we have explored other options, specifically using the
+[Triton][triton] compiler to author kernels. The `torch.compile` system introduced in PyTorch
+2.0 uses Triton to generate GPU kernels. We now have prototype implementations for `DSD` and `SDD`
+matrix multiplication using BSR layout available in PyTorch 2.1.
+
+## Benchmarks
+
+We have evaluated the performance of new kernels. The performance is measured by speedup compared to
+the dense implementation. Where ever the functionality exists in both pytorch 2.1 and 2.0 we show
+the metrics for both versions.  All experiments are performed using the `cuda` device type on a
+system with a single NVIDIA A100-80G, and CUDA 11.8. 
+
+### `sampled_addmm` (SDD)
+In the case of this operation a dense equivalent would involve the composition between matrix
+multiplication and a masking operation.  Here, we have chose for the dense baseline a normal matrix
+multiplication without masking. Since in practice the masking operation may not be used if sparse
+tensors are not also in use, for example when computing gradients. This gives us a harder target to
```

Comment:
```suggestion
multiplication without masking since, in practice, the masking operation may not be used if the inputs are not 
sparse, for example, when computing gradients. This gives us a harder target to
```